Don't clear PHP buffer when an error occurs with debug=true

lyrixx opened the pull request:
fixes #3058

lyrixx committed:
Don't clear PHP buffer when an error occurs with debug=true

lyrixx commented on src/Extension/DebugExtension.php:
```diff
@@ -54,7 +54,7 @@ function twig_var_dump(Environment $env, $context, array $vars = [])
     }
 
-    ob_start(function () { return ''; });
+    ob_start();
```
In that case, we are already in debug mode

stof commented on src/Node/MacroNode.php:
```diff
@@ -104,7 +104,13 @@ public function compile(Compiler $compiler)
             ->write("\$blocks = [];\n\n")
-            ->write("ob_start(function () { return ''; });\n")
+        ;
+        if ($compiler->getEnvironment()->isDebug()) {
```
Do we take the debug mode into account in the cache key ? If no, we would need to make that check at runtime rather than at compile time.

lyrixx replied:
I don't recall ... :/

stof replied:
checked it. The Environment puts it in its option hash, so that's fine.

fabpot commented:
@lyrixx Can you add a test to check that we don't leak output on non-debug mode and that we have the output on debug mode? Also, can you add a note in the CHANGELOG?

lyrixx commented:
@fabpot Done. That was not easy because PHPUnit buffer output + has an error handler. So the simplest thing to do was to use an external script

stof commented on test/Twig/Tests/ErrorTest.php:
```diff
@@ -216,3 +238,20 @@ public function bar()
 }
+
+
+class BrokenExtension extends \Twig\Extension\AbstractExtension
```
this one looks unused

lyrixx replied:
Oups, thanks.

fabpot commented:
@lyrixx Tests do not pass for 5.4 :)

lyrixx commented:
Woho, PHP 5.4. Thanks for report. I fixed that in a very clean way :joy:

fabpot commented:
Yay for PHP 5.4... but some other PHP versions are now broken :joy:

fabpot commented on CHANGELOG:
```diff
@@ -1,6 +1,6 @@
 * 1.42.2 (2019-XX-XX)
 
- * n/a
+ * Display partial output (PHP buffer) when an eeor occurs in debug mode
```
typo: error

fabpot commented on composer.json:
```diff
@@ -30,6 +30,7 @@
     "require-dev": {
         "symfony/phpunit-bridge": "^3.4.19|^4.1.8|^5.0",
         "symfony/debug": "^2.7",
+        "symfony/process": "^2.7|3.4|^4.0|^5.0",
```
typo: `^3.4`

fabpot commented on test/Twig/Tests/Fixtures/errors/leak-output.php:
```diff
@@ -0,0 +1,29 @@
+
+    public function broken()
+    {
+        die('OUPS');
```
I would prefer the English version `OOPS` 😂

fabpot commented on test/Twig/Tests/Fixtures/errors/leak-output.php:
```diff
@@ -0,0 +1,29 @@
+    }
+}
+
+$loader = new Twig_Loader_Array([
```
You need to use the namespaced classes (same below).

lyrixx commented:
I stop using symfony/process for using plain `exec`. Simpler to manage! And now it's all green :green_heart: 
I also addressed your comments.
Thanks.

fabpot commented:
Thank you @lyrixx.

stof commented on test/Twig/Tests/ErrorTest.php:
```diff
@@ -207,6 +207,20 @@ public function getErroredTemplates()
+
+    public function testTwigLeakOutputInDebugMode()
+    {
+        $output = exec(sprintf('%s %s debug', \PHP_BINARY, __DIR__.'/Fixtures/errors/leak-output.php'));
```
you need to escape the argument for the path (`__DIR__` might have spaces in it)

lyrixx replied:
Good catch. https://github.com/twigphp/Twig/pull/3062

lyrixx commented:
@fabpot Could you do a release with this patch? This is quite annoying.  Thanks

fabpot commented:
@lyrixx New releases out just for you :)

lyrixx commented:
Thanks :)
